### PR TITLE
btc: use blockfilter when requesting address info

### PR DESCRIPTION
**Pull request by ixv**
Now when we get updated about a new block, only request address info for addresses that match that block's blockfilter. If there's a gap of multiple blocks, we need to go request blockfilter information for those blocks specificially. 

This depends on this PR on the urbit-bitcoin-rpc repo, which allows you to make those requests for specific block information: https://github.com/timlucmiptev/urbit-bitcoin-rpc/pull/26

### CODE REVIEW

**Comment by timlucmiptev** [comment]
Looks good, just two comments/questions:

1. Do we want to expose bip158 via bitcoin.hoon? I split up the various BIP functionality just for modularity, but it doesn't seem like a relevant detail to app developers.

2. Do you want to have a `gap` size past which you just check addresses in the normal way? If a wallet was offline a long time and missed tons of blocks, the fetching of blockfilters would take longer than just checking addresses, since blockfilters aren't *that* small (~10% of total block size iirc).

**Comment by ixv** [comment]
1. Not sure I understand, do you mean doing `/+  bip-b158` in `/lib/bitcoin.hoon` rather than in  `/app/btc-wallet.hoon`, or did you have something else in mind?

2. Yeah that's a good idea, what would you suggest is a good gap size?

**Comment by timlucmiptev** [comment]
1. Yeah, include it in `lib/bitcoin.hoon`. If we only use 1 or 2 of its functions externally (I think that's the case?), you could even wrap just those in arms, which seems best now that I think about it.

2. I just checked the blockfilter for the latest block: it was ~32K bytes on a 1.5M block, so about 2% of the size. In terms of what we're good sending over Ames, I think about 50 blocks would be good here? 1.5M on a reconnect seems reasonable, but if you feel it's too much, use that ~30KB number to adjust as appropriate.

**Comment by timlucmiptev** [comment]
If `all-match` is the only thing we use externally from BIP158, I would only expose that and `match` as public arms through `bitcoin.hoon`.

Other than that, looks good. Might want a review from @tacryt-socryp  on the provider kick/re-sub.

File: pkg/arvo/app/btc-wallet.hoon [hunk 1]
```diff
@@ -704,102 +690,148 @@
...
-    `(poke-provider [%tx-info txid])
-  --
+  =?  cards  ?|(!connected.p (gth gap 0))
+    ;:  weld  cards
+      (retry-poym network)
+      (retry-txs network)
+      (retry-scans network)
+    ==
+  =?  cards  ?&(?=(^ blockhash) ?=(^ blockfilter) (gth gap 0))
+    ;:  weld  cards
```
Comment (tacryt-socryp) [inline pkg/arvo/app/btc-wallet.hoon:727]: Use regular form with %+ when only using two items.

File: pkg/arvo/app/btc-wallet.hoon [hunk 2]
```diff
@@ -704,102 +690,148 @@
...
+      (retry-filtered-addrs network u.blockhash u.blockfilter)
+    ==
+  =?  cards  (gth gap 50)
+    ;:  weld  cards
```
Comment (tacryt-socryp) [inline pkg/arvo/app/btc-wallet.hoon:731]: Regular form

File: pkg/arvo/app/btc-wallet.hoon [hunk 3]
```diff
@@ -704,102 +690,148 @@
...
+  |=  =txid
+  (poke-provider [%tx-info txid])
+::
+++  retry-scans
+  |=  =network
+  ^-  (list card)
+  %-  zing
+  %+  murn  ~(tap by scans)
+  |=  [[=xpub:bc =chyg] =batch]
+  ?.  =(network network:(~(got by walts) xpub))  ~
```
Comment (tacryt-socryp) [inline pkg/arvo/app/btc-wallet.hoon:758]: Crashing here in the got by seems like a bad idea. Let's make sure we produce ~ instead

**Review by tacryt-socryp — CHANGES_REQUESTED** [review]
I don't see anything here to handle a negative watch ack. We want to ensure we don't cause a resubscribe loop in that case. Otherwise, small style nits

**Comment by ixv** [comment]
Not sure I follow, if we don't want to do anything in the negative watch-ack case, why isn't the default-agent handling of this sufficient?

**Comment by tacryt-socryp** [comment]
Line 729 won't compile due to using a gap inside of the (weld ... ) statement. As for the watch nack, I think you're right, but we need a more semantic error message for the crash, likely with a `~|`

**Comment by ixv** [comment]
Alright done, also found and fixed another issue with it

**Review by tacryt-socryp — APPROVED** [review]

**Pull request merged**